### DISCUSSION TITLE
Migerate Firebase Authentication (SCRYPT) to zitadel

### DISCUSSION THREAD

**Discussion by approached** [zitadel/zitadel#6921] [Q&A]
Hi @ all

How can i migrate Firebase Authentication (SCRYPT) to zitadel?

It used SCRYPT as hashed passwords:
https://firebaseopensource.com/projects/firebase/scrypt/#other_languages

**Comment by muhlemmer** [comment]
We support scrypt imports in the Modular Crypt Format, which looks like:

> Scrypt uses standard raw Base64 encoding (no padding) for the salt and hash.
> The resulting Modular Crypt Format string looks as follows:
> 
> ```
> $scrypt$ln=16,r=8,p=1$cmFuZG9tc2FsdGlzaGFyZA$Rh+NnJNo1I6nRwaNqbDm6kmADswD1+7FTKZ7Ln9D8nQ
>   (1)        (2)              (3)                              (4)
> ```
> 
> 1. The identifier is always `scrypt`.
> 2. Cost parameters:
>    1. `ln` is the exponential cost parameter for memory and CPU - `16` in this example.
>    2. `r` is the block size for optimal performance of the CPU architecture - `8` in this example.
>    3. `p` is to indicate parallelism - `1` in this example.
> 3. Base64-encoded salt
> 4. Base64-encoded Scrypt hash output of the password and salt combined.

The above is from a documentation update I'm working on. In order to import scrypt, you will need the salt and hash without padding. (you can discard any `=` padding chars) and build the above string with the same parameters used when creating the hash.

I read the firebase documentation and the python implementation. After the key-derivation function is ran, they do an additional step of AES-encrypting the `SigningKey` and store that result. We currently do not support that. That seems to be a Firebase customization to the Scrypt algorithm. We use the [scrypt Go package](https://pkg.go.dev/golang.org/x/crypto/scrypt)  and claim compatibility with[ Python's passlib](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.scrypt.html)

It should not be difficult to add the firebase variant to our [passwap library](https://github.com/zitadel/passwap) (and thus zitadel), if there is enough demand either by community or commercial customer.

**Comment by approached** [comment]
Thank you very much for your answer. So I have two options:

1. wait and pray that someone expands the library.
2. or manually start the migration process at login and create the user with his password in zitadel

**Comment by muhlemmer** [comment]
> wait and pray that someone expands the library.

If you want to use the free or open source version, you depend on community support yes.  So a PR is always welcome as well.

We have also [commercial support options](https://zitadel.com/pricing) and we prioritize features for paying users. If you need help migrating from firebase to zitadel, please reach out to hi@zitadel.com and I'm sure there is something that can be worked out.

**Reply by muhlemmer** [reply]
More info also here: https://zitadel.com/docs/legal/policies/feature-development-policy
